### PR TITLE
Fix Health Check to support Enhanced Recurring Donations

### PR DESCRIPTION
**Pull request by force2b**
# Critical Changes

# Changes

- Health Check for Recurring Donations has been updated to more accurately validate the Open Ended Status, Installment Period, Schedule Type picklist fields, and to support the upcoming release of Enhanced Recurring Donations.

# Issues Closed

# Community Ideas Delivered

# New Metadata

Apex Classes:
- RD_Constants
- STG_PanelRDHealthCheck
- STG_PanelRDHealthCheck_TEST

Custom Labels:
- healthDetailsInvalidPicklistValue
- healthDetailsMissingRequiredPicklistValue
- healthSolutionMissingPicklistValue
- healthSolutionUnsupportedPicklistValue

# Deleted Metadata

### CODE REVIEW

File: src/labels/CustomLabels.labels [hunk 1]
```diff
@@ -4883,6 +4883,14 @@ To check your GAU Allocation settings, go to Donations | GAU Allocations.</value
         <shortDescription>healthDetailsInvalidRDCustomPeriodPIcklist</shortDescription>
         <value>Recurring Donation Installment Period picklist value {0} is not a valid Custom Installment Period.</value>
     </labels>
+    <labels>
```
Comment (force2b) [inline src/labels/CustomLabels.labels]: @sam-knox Would you mind reviewing these new custom labels? These are used to display recurring donation health check errors for either unsupported picklist values or missing required picklist values. There are existing labels for some of the other recurring donations health check operations that are unchanged. Thanks.

File: src/classes/STG_PanelHealthCheck_CTRL.cls [hunk 2]
```diff
@@ -1,10 +1,10 @@
 /*
     Copyright (c) 2014 Salesforce.org
     All rights reserved.
-    
+
```
Comment (force2b) [inline src/classes/STG_PanelHealthCheck_CTRL.cls]: _Changes in this file are 99% whitespace related. The only actual code change is line 176._

File: src/classes/RD_Constants.cls [hunk 3]
```diff
@@ -0,0 +1,64 @@
+/*
```
Comment (force2b) [inline src/classes/RD_Constants.cls]: _This file is a direct copy from the #4997 branch._

File: src/labels/CustomLabels.labels (hunk 1 above)
Comment (sam-knox) [reply src/labels/CustomLabels.labels]: @force2b - Will it be clear to the user what field the error is on? Just saying that a picklist value isn't supported might lead to the question: Which picklist?

Comment (force2b) [reply src/labels/CustomLabels.labels]: Good question. The related _solution_ label will have both the picklist value (`{0}`) and the label of the picklist field (`{1}`) to fix.

**Comment by force2b** [comment]
@sam-knox Thanks. Changes made.

**Comment by force2b** [comment]
**lurch: attach W-038187

**Comment by LurchTheButler** [comment]
Tracking <a href="https://foundation.my.salesforce.com/a2x1E000001HptnQAC" target="_blank">W-038187</a>

File: src/classes/STG_PanelRDHealthCheck_TEST.cls [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+     * @description Validates that Enhanced Recurring Donations is configured properly in the org.
+     * Because RD1 is the default for NPSP, there are some settings that will fail the health check
+     * such as for the Quarterly installment period. Situations like this are validated for in the
+     * method.
+     */
+    @IsTest
+    private static void shouldValidateEnhancedRecurringDonationsConfig() {
+        RD2_EnablementService_TEST.setRecurringDonations2Enabled();
+
+        npe03__Recurring_Donations_Settings__c rds = UTIL_CustomSettingsFacade.getRecurringDonationsSettingsForTest(
```
Comment (RadSFDeveloper) [inline src/classes/STG_PanelRDHealthCheck_TEST.cls]: Settings creation can be reused as per [RD2_Settings_TEST.setUpConfiguration()](https://github.com/SalesforceFoundation/NPSP/blob/master/src/classes/RD2_Settings_TEST.cls#L191) while extra fields (if any) can be set in the input Map.

File: src/classes/STG_PanelRDHealthCheck_TEST.cls [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+
+        String results = getValidationResults(ctrl);
+        System.assertEquals(0, ctrl.listDR.size(),
+            'There should zero errors in this picklist validation\n' + results
+        );
+
+        // Add an invalid "required" picklist to this list and validate that the required
+        // validation logs the error.
+        // =============================================================
+        supportedInstallmentPeriods.add(BAD_PICKLIST_VALUE);
```
Comment (RadSFDeveloper) [inline src/classes/STG_PanelRDHealthCheck_TEST.cls]: This seems beginning of another concept verification, should it be extracted into another unit test?

File: src/classes/STG_PanelRDHealthCheck_TEST.cls [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+        String results = getValidationResults(ctrl);
+        System.assertEquals(1, ctrl.listDR.size(),
+            'There should one result in the installment validation.\n' + results
+        );
+        System.assertEquals(STG_PanelHealthCheck_CTRL.statusSuccess, ctrl.listDR[0].strStatus,
+            'The result status should have been "' + STG_PanelHealthCheck_CTRL.statusSuccess + '".\n' + results
+        );
+        ctrl.listDR.clear();
+
+        // =============================================================
```
Comment (RadSFDeveloper) [inline src/classes/STG_PanelRDHealthCheck_TEST.cls]: This seems beginning of another concept, should it be extracted into another unit test?

File: src/classes/STG_PanelRDHealthCheck_TEST.cls [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+        System.assert(ctrl.listDR[0].strDetails.contains(mockRecordTypeId),
+            'The result status details should contained the bad record type id of "' + mockRecordTypeId + '".\n' + results
+        );
+    }
+
+    // ==================================================================================
+    // HELPER METHODS
+    // ==================================================================================
+
+    private static String getValidationResults(STG_PanelHealthCheck_CTRL ctrl) {
```
Comment (RadSFDeveloper) [inline src/classes/STG_PanelRDHealthCheck_TEST.cls]: ApexDoc missing.

File: src/classes/STG_PanelRDHealthCheck_TEST.cls [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+        STG_PanelRDHealthCheck healthCheck = new STG_PanelRDHealthCheck(ctrl);
+
+        healthCheck.validateOpportunityStage(PLEDGED_STAGE, false, false);
+
+        String results = getValidationResults(ctrl);
+        System.assertEquals(0, ctrl.listDR.size(),
+            'There should zero errors in the record type validation\n' + results
+        );
+
+        healthCheck.validateOpportunityStage(PLEDGED_STAGE, true, true);
```
Comment (RadSFDeveloper) [inline src/classes/STG_PanelRDHealthCheck_TEST.cls]: This seems beginning of another concept, should it be extracted into another unit test?
Also, I think we should name methods to convey what is being validated without providing input Booleans. For example, we can add into the `STG_PanelRDHealthCheck` and verify those calls from tests? If not let's provide vars to store Boolean input args?
```

    /**
     * @description Validates specified stage name exists as an open stage and it is active
     * @param oppStageName Opportunity Stage Name
     * @return void
     */
    private void validateOpenOpportunityStage(String oppStageName) {
        validateOpportunityStage(oppStageName, false, false);
    }

    /**
     * @description Validates specified stage name exists as a Closed Lost stage and it is active
     * @param oppStageName Opportunity Stage Name
     * @return void
     */
    private void validateClosedLostOpportunityStage(String oppStageName) {
        validateOpportunityStage(oppStageName, false, true);
    }
```

File: src/classes/STG_PanelRDHealthCheck.cls [hunk 9]
```diff
@@ -0,0 +1,430 @@
...
+    ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+    POSSIBILITY OF SUCH DAMAGE.
+*/
+/**
+* @author Salesforce.org
+* @date 2020
+* @group Settings
+* @description Recurring Donations Health Check
+*/
+public class STG_PanelRDHealthCheck {
```
[the remark this answers is not in the document]
Comment (force2b) [reply src/classes/STG_PanelRDHealthCheck.cls]: I'd go `inherited sharing' so it uses what the calling class implements.

File: src/classes/STG_PanelRDHealthCheck.cls [hunk 10]
```diff
@@ -0,0 +1,430 @@
...
+        Set<String> supportedScheduleTypes = new Set<String>();
+        supportedScheduleTypes.add(RD_Constants.SCHEDULE_TYPE_MULTIPLY_BY);
+        supportedScheduleTypes.add(RD_Constants.SCHEDULE_TYPE_DIVIDE_BY);
+
+        validateSupportedPicklistValues(supportedScheduleTypes, 'npe03__Schedule_Type__c',
+            System.Label.healthDetailsInvalidPicklistValue, System.Label.healthSolutionUnsupportedPicklistValue);
+        validateRequiredPicklistValues(supportedScheduleTypes, 'npe03__Schedule_Type__c',
+            System.Label.healthDetailsMissingRequiredPicklistValue, System.Label.healthSolutionMissingPicklistValue);
+
+        if (hasValidationErrors()) {
```
[the remark this answers is not in the document]
Comment (force2b) [reply src/classes/STG_PanelRDHealthCheck.cls]: Good catch. That should have been `!hasValidationErrors()`

File: src/classes/STG_PanelRDHealthCheck.cls [hunk 11]
```diff
@@ -0,0 +1,430 @@
...
+
+        // =============================================================
+        // Verify Status Values
+        // =============================================================
+        Set<String> supportedStatus = new Set<String>();
+        supportedStatus.add(RD2_Constants.STATUS_ACTIVE);
+        supportedStatus.add(RD2_Constants.STATUS_CLOSED);
+        supportedStatus.add(RD2_Constants.STATUS_LAPSED);
+
+        validateSupportedPicklistValues(supportedStatus, UTIL_Namespace.StrAllNSPrefix('Status__c'),
```
[the remark this answers is not in the document]
Comment (force2b) [reply src/classes/STG_PanelRDHealthCheck.cls]: True. That works.

File: src/classes/STG_PanelRDHealthCheck.cls [hunk 12]
```diff
@@ -0,0 +1,430 @@
...
+                    )
+                );
+            }
+        }
+    }
+
+    /**
+     * @description Validate that the specified Record Type Api Name exists as Active and defined as
+     * Closed and/or Won as appropriate
+     * @param oppStageName Record Type Api Name (DeveloperName NOT label)
```
[the remark this answers is not in the document]
Comment (force2b) [reply src/classes/STG_PanelRDHealthCheck.cls]: The apex doc is correct. The value of the custom label would have to represent the api name of the stage. This is all part of my dislike for the use of custom labels for picklist values. It was a mistake for those to ever be used, so now we live with the decision until we eventually move them into custom settings fields - which is what we did for the Installment Stage Name in RD2.

File: src/classes/STG_PanelRDHealthCheck.cls [hunk 13]
```diff
@@ -0,0 +1,480 @@
...
+        private final String statusError = STG_PanelHealthCheck_CTRL.statusError;
+        private final String statusSuccess = STG_PanelHealthCheck_CTRL.statusSuccess;
+
+        private STG_PanelHealthCheck_CTRL ctrl;
+        private String scope;
+
+        public HealthCheckResult(STG_PanelHealthCheck_CTRL ctrl) {
+            this.ctrl = ctrl;
+        }
+
```
Comment (RadSFDeveloper) [inline src/classes/STG_PanelRDHealthCheck.cls]: Missing ApexDoc on the new inner class methods.

Comment (force2b) [reply src/classes/STG_PanelRDHealthCheck.cls]: Thanks. Done.

**Comment by rponti-sforg** [comment]
QE testing for this PR and corresponding WI is now complete.  Based on PR comments / feedback, WI description, acceptance criteria and test charter discussion and feedback, tests have passed.

For further testing details, see:
https://foundation.lightning.force.com/lightning/r/agf__ADM_Work__c/a2x1E000001HptnQAC/view
https://salesforce.quip.com/P4BAA6hj53W6

Merging PR.  Will set WI status to Closed once post merge PR builds are complete.

**Pull request merged**